src/Bicep.Core/Registry/ModuleDispatcher.cs

```diff
@@ -85,14 +89,16 @@ public ResultWithDiagnostic<ArtifactReference> TryGetModuleReference(string refe
             }
         }
 
-        public ResultWithDiagnostic<ArtifactReference> TryGetModuleReference(IArtifactReferenceSyntax moduleDeclaration, Uri parentModuleUri)
+        public ResultWithDiagnostic<ArtifactReference> TryGetArtifactReference(IArtifactReferenceSyntax artifactDeclaration, Uri parentModuleUri)
         {
-            if (!SyntaxHelper.TryGetForeignTemplatePath(moduleDeclaration).IsSuccess(out var moduleReferenceString, out var failureBuilder))
+            if (!SyntaxHelper.TryGetForeignTemplatePath(artifactDeclaration).IsSuccess(out var artifactReferenceString, out var failureBuilder))
             {
                 return new(failureBuilder);
             }
 
-            return this.TryGetModuleReference(moduleReferenceString, parentModuleUri);
+            var artifactType = (artifactDeclaration is ProviderDeclarationSyntax) ? ArtifactType.Provider : ArtifactType.Module;
```

Comment:
I think this is OK as a one-off. If you find yourself doing this pattern in other places, then it might be worth making `GetArtifactType()` a method on `IArtifactReferenceSyntax` to avoid tying code that's dependent on abstractions (interfaces) to concrete implementations (subclasses).